File: common_codes/a8_statcalculator.m
function Res=a8_statcalculator(WQIs)
exp=WQIs(:,1); pred=WQIs(:,2:end);
AE=abs((exp-pred));
MAE=mean(AE,1,"default");

% APE=(AE./exp)*100;
% MAPE=mean(APE,1,"omitmissing");

SE=AE.^2;
MSE=mean(SE,1,"default");
RMSE=MSE.^0.5;

% SSE=sum(SE,1,"default");
% SST=sum((exp-mean(exp,1,"default")).^2,1,"default");
% R2=1-(SSE./SST);

Res=[MAE;RMSE];
Res=array2table(Res,"RowNames",{'MAE','RMSE'},"VariableNames",{'LWQI','AWQI','RWQI'});

File: python_code/common_codes/a8_statcalculator.py
import numpy as np
import pandas as pd


def a8_statcalculator(WQIs):
    exp = WQIs[:, 0]
    pred = WQIs[:, 1:]
    AE = np.abs(exp - pred[:, None] if pred.ndim == 1 else exp[:, None] - pred)
    AE = np.abs(exp[:, np.newaxis] - pred)
    MAE = np.mean(AE, axis=0)
    SE = AE ** 2
    MSE = np.mean(SE, axis=0)
    RMSE = MSE ** 0.5

    Res = np.vstack([MAE, RMSE])
    Res = pd.DataFrame(Res, index=['MAE', 'RMSE'], columns=['LWQI', 'AWQI', 'RWQI'])
    return Res